File: library/src/main/java/com/xwray/groupie/AsyncDiffUtil.java
package com.xwray.groupie;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListUpdateCallback;

import java.util.List;

/**
 * A wrapper around {@link DiffUtil} that calculates diff in a background thread
 */
class AsyncDiffUtil {

    interface Callback extends ListUpdateCallback {
        /**
         * Called on the main thread before DiffUtil dispatches the result
         */
        @MainThread
        void onDispatchAsyncResult(List<Group> mergedGroups);
    }

    final Callback asyncDiffUtilCallback;
    private int maxScheduledGeneration;

    AsyncDiffUtil(@NonNull Callback callback) {
        this.asyncDiffUtilCallback = callback;
    }

    int getMaxScheduledGeneration() {
        return maxScheduledGeneration;
    }

    void calculateDiff(@NonNull List<? extends Group> oldGroups,
                       @NonNull List<? extends Group> newGroups,
                       @Nullable final OnAsyncUpdateListener onAsyncUpdateListener,
                       boolean detectMoves) {
        // incrementing generation means any currently-running diffs are discarded when they finish
        final int runGeneration = ++maxScheduledGeneration;
        final DiffCallback diffUtilCallback = new DiffCallback(oldGroups, newGroups);
        new DiffTask(this, diffUtilCallback, runGeneration, detectMoves, onAsyncUpdateListener).execute();
    }
}

File: library-kotlin/src/main/java/com/wray/groupiekotlin/AsyncDiffUtil.kt
package com.wray.groupiekotlin

import androidx.annotation.MainThread
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListUpdateCallback

/**
 * Listener that provides a callback when
 * async update methods complete.
 */
interface OnAsyncUpdateListener {
    fun onUpdateComplete()
}

/**
 * A wrapper around [DiffUtil] that calculates diff in a background thread
 */
internal class AsyncDiffUtil(val asyncDiffUtilCallback: Callback) {
    var maxScheduledGeneration: Int = 0
        private set

    lateinit var groups: Collection<Group>
        private set

    internal interface Callback : ListUpdateCallback {
        /**
         * Called on the main thread before DiffUtil dispatches the result
         */
        @MainThread
        fun onDispatchAsyncResult(newGroups: Collection<Group>)
    }

    fun calculateDiff(newGroups: Collection<Group>,
                      diffUtilCallback: DiffUtil.Callback,
                      onAsyncUpdateListener: OnAsyncUpdateListener?,
                      detectMoves: Boolean) {
        groups = newGroups
        // incrementing generation means any currently-running diffs are discarded when they finish
        val runGeneration = ++maxScheduledGeneration
        DiffTask(this, diffUtilCallback, runGeneration, detectMoves, onAsyncUpdateListener).execute()
    }
}
